File: tks/ApplicationCore/ApplicationServices/src/main/kotlin/services/RentService.java
package services;

import domain.exceptions.*;
import ports.primary.IRentService;
import ports.primary.ResourceRentCommandPort;
import ports.secondary.RentPersistencePort;
import ports.secondary.RentSearchPort;
import domain.model.values.AccessionNumber;
import domain.model.values.Email;
import ports.secondary.ResourceSearchPort;
import ports.secondary.UserSearchPort;
import domain.model.Rent;

import java.util.Date;
import java.util.UUID;

public class RentService implements IRentService {

   private final RentPersistencePort rentPersistencePort;
   private final RentSearchPort rentSearchPort;
   private final UserSearchPort userSearchPort;
   private final ResourceSearchPort resourceSearchPort;

    public RentService(RentPersistencePort rentPersistencePort, RentSearchPort rentSearchPort, UserSearchPort userSearchPort, ResourceSearchPort resourceSearchPort) {
        this.rentPersistencePort = rentPersistencePort;
        this.rentSearchPort = rentSearchPort;
        this.userSearchPort = userSearchPort;
        this.resourceSearchPort = resourceSearchPort;
    }

    public void rent(Email email, AccessionNumber resourceId) throws UserNotFoundException, ResourceNotFoundException, UserNotActiveException, ResourceAlreadyRentException {
        var user = userSearchPort.findByEmail(email);
        if(user == null) throw new UserNotFoundException();
        var resource = resourceSearchPort.findById(resourceId);
        if(resource == null) throw new ResourceNotFoundException();
        if(!user.getActive()) throw new UserNotActiveException();
        var existingRent = rentSearchPort.findActiveByResourceId(resourceId);
        if(existingRent != null) throw new ResourceAlreadyRentException();
        var rent = new Rent(UUID.randomUUID(), new Date(), null, user.getEmail(), resourceId);
        rentPersistencePort.save(rent);
    }

    public void returnResource(Email email, AccessionNumber resourceId) throws UserNotFoundException, ResourceNotFoundException, ResourceNotRentException, InvalidUserException {
        var user = userSearchPort.findByEmail(email);
        if(user == null) throw new UserNotFoundException();
        var resource = resourceSearchPort.findById(resourceId);
        if(resource == null) throw new ResourceNotFoundException();
        var rent = rentSearchPort.findActiveByResourceId(resourceId);
        if(rent == null) throw new ResourceNotRentException();
        if(!rent.getUserEmail().equals(user.getEmail())) throw new InvalidUserException();
        rent.setEndDate(new Date());
        rentPersistencePort.save(rent);
    }
}

File: tks/ApplicationCore/ApplicationServices/src/main/kotlin/application/services/RentService.kt
package application.services

import domain.exceptions.*
import domain.model.context.rents.Rent
import domain.model.values.AccessionNumber
import domain.model.values.Email
import ports.primary.combined.IRentService
import ports.secondary.*
import java.util.*

open class RentService(
    private val rentPersistencePort: RentPersistencePort,
    private val rentSearchPort: RentSearchPort,
    private val clientSearchPort: ClientSearchPort,
    private val productSearchPort: ProductSearchPort
) : IRentService {

    @Throws(
        UserNotFoundException::class,
        ResourceNotFoundException::class,
        UserNotActiveException::class,
        ResourceAlreadyRentException::class
    )
    override fun rent(email: Email, resourceId: AccessionNumber): UUID {
        val (userEmail, active) = clientSearchPort.findByEmail(email) ?: throw UserNotFoundException()
        productSearchPort.findByAccessionNumber(resourceId) ?: throw ResourceNotFoundException()
        if (!active) throw UserNotActiveException()
        val existingRent = rentSearchPort.findActiveByResourceId(resourceId)
        if (existingRent != null) throw ResourceAlreadyRentException()
        val rent = Rent(UUID.randomUUID(), Date(), null, userEmail, resourceId)
        rentPersistencePort.save(rent)
        return rent.id!!
    }

    @Throws(
        UserNotFoundException::class,
        ResourceNotFoundException::class,
        ResourceNotRentException::class,
        InvalidUserException::class
    )
    override fun returnResource(email: Email, resourceId: AccessionNumber) {
        val (userEmail) = clientSearchPort.findByEmail(email) ?: throw UserNotFoundException()
        productSearchPort.findByAccessionNumber(resourceId) ?: throw ResourceNotFoundException()
        val rent = rentSearchPort.findActiveByResourceId(resourceId) ?: throw ResourceNotRentException()
        if (rent.userEmail != userEmail) throw InvalidUserException()
        rent.endDate = Date()
        rentPersistencePort.save(rent)
    }

    @Throws(RentNotFoundException::class)
    override fun getDetails(id: UUID): Rent {
        return rentSearchPort.getById(id) ?: throw RentNotFoundException();
    }
}